### PR TITLE
Add support for common_io on ESP32

### PR DESCRIPTION
**Pull request by shubhamkulkarni97**
Description
-----------
This PR intends to support common_io (device driver layer) on ESP32.

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [ ] My code is Linted.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by teuteuguy** [comment]
Started looking at this PR for using common-io for I2C.

ESP has 2x I2C ports with 2 sets of pinouts for SDA and SCL.
Currently setup is made using pins 18 and 19.

How to setup pins 21 and 22 ?

**Comment by shubhamkulkarni97** [comment]
> Started looking at this PR for using common-io for I2C.
> 
> ESP has 2x I2C ports with 2 sets of pinouts for SDA and SCL.
> Currently setup is made using pins 18 and 19.
> 
> How to setup pins 21 and 22 ?

Hi @teuteuguy ,

You can specify pin mappings in `board_gpio.h` (https://github.com/aws/amazon-freertos/blob/41abade73c17ff638df2651405dad091248272ea/vendors/espressif/boards/esp32/ports/common_io/include/board_gpio.h#L64).

**Comment by teuteuguy** [comment]
OK. @shubhamkulkarni97 you were right, for some reason I did not have the latest code.
I've modified the board_gpio to reflect my specific pins.

Problem 1:
Now still trying to work with the I2C I'm struggling to figure out how to get it to work.
Here is an example "sample" way of reading a byte from I2C from pure espressif code:

i2c_cmd_handle_t cmd = i2c_cmd_link_create();
i2c_master_start(cmd);
i2c_master_write_byte(cmd, (device_addr << 1) | I2C_MASTER_WRITE, ACK_CHECK_EN);
i2c_master_write_byte(cmd, reg_addr, ACK_CHECK_EN);
i2c_master_start(cmd);
i2c_master_write_byte(cmd, (device_addr << 1) | I2C_MASTER_READ, ACK_CHECK_EN);
i2c_master_read_byte(cmd, rd_buff + length - 1, NACK_VAL);
i2c_master_stop(cmd);
e = i2c_master_cmd_begin(i2c_port_num, cmd, 250 / portTICK_RATE_MS);
i2c_cmd_link_delete(cmd);

Now the problem with the common_io implementation is that all of this needs to be split with the different calls required to io_ctl and the flow would then become something like (as per header example):
// Set slave address.
e = iot_i2c_ioctl( pxI2CPeripheral, eI2CSetSlaveAddr, &address );
// Write the register address as single byte, in a transaction.
e = iot_i2c_write_sync( pxI2CPeripheral, reg, 1 );
// Read length bytes of data to allocated buffer, in a transaction.
e = iot_i2c_read_sync( pxI2CPeripheral, values, 1 );

However how does this work ?

e = iot_i2c_ioctl( pxI2CPeripheral, eI2CSetSlaveAddr, &address ); gets implemented as:
        i2c_cmd_handle_t cmd = i2c_cmd_link_create();
        ret = i2c_master_start(cmd);
        ret |= i2c_master_write_byte(cmd, (*(uint8_t *)pvBuffer) << 1 | I2C_MASTER_WRITE, ACK_CHECK_EN);
        i2c_cmd_link_delete(cmd);

e = iot_i2c_write_sync( pxI2CPeripheral, reg, 1 ); gets implemented as:
        i2c_cmd_handle_t cmd = i2c_cmd_link_create();
        ret = i2c_master_write(cmd, src_buf, xBytes, ACK_CHECK_EN);
        ret |= i2c_master_cmd_begin(i2c_port_num, cmd, 1000 / portTICK_RATE_MS );
        i2c_cmd_link_delete(cmd);

e = iot_i2c_read_sync( pxI2CPeripheral, values, 1 ); gets implemented as:
        i2c_cmd_handle_t cmd = i2c_cmd_link_create();
        ret |= i2c_master_read_byte(cmd, (uint8_t *)(src_buf + xBytes - 1), NACK_VAL);
        ret |= i2c_master_stop(cmd);
        ret |= i2c_master_cmd_begin(i2c_port_num, cmd, 1000 / portTICK_RATE_MS);
        i2c_cmd_link_delete(cmd);


And so the following is missing from the flow:
i2c_master_write_byte(cmd, (device_addr << 1) | I2C_MASTER_READ, ACK_CHECK_EN);

And also wonder if i2c_master_start(cmd); isn't missing in the iot_i2c_write_sync and iot_i2c_read_sync functions.

Problem 2: 
There seems to be a bug in the iot_i2c_read_sync function with the following "!!" that I believe should only be "!" when checking for !iot_i2c_handler->is_slave_addr_write_set

Problem 3:
We can see that some of the capabilities of common_io allow for: eI2CGetTxNoOfbytes, however from my understanding ESP32 does not have this kind of capability. As such, if someone is trying to create a portable library with common_io in mind, how would he go about knowing if eI2CGetTxNoOfbytes is supported or not and thus knowing to implement the check or not ?

**Comment by teuteuguy** [comment]
Forgot to add that I've been trying the above code and comparing direct implementation and common_io to read 1 register off of an I2C device connected to my ESP32 and the common_io version does not work. 
I believe because of the following missing part.

Also wanting to add, that am not sure how the io_ctl for eI2CSetSlaveAddr works in specifying/setting the following 2 flags separately:
    bool is_slave_addr_write_set;
    bool is_slave_addr_read_set;

**Comment by shubhamkulkarni97** [comment]
Hi @teuteuguy ,

> And so the following is missing from the flow:
> i2c_master_write_byte(cmd, (device_addr << 1) | I2C_MASTER_READ, ACK_CHECK_EN);
>

While reading data from I2C slave device function call to set slave address is missing. We are working on this, to get the sequence of commands in sync with the pure I2C code in ESP-IDF. 
> And also wonder if i2c_master_start(cmd); isn't missing in the iot_i2c_write_sync and iot_i2c_read_sync functions.
> 
Start command is sent through `iot_i2c_ioctl` while setting slave address (https://github.com/shubhamkulkarni97/amazon-freertos/blob/0c6b271f89ab70ce0238d26a49a54bd2411ecdf2/vendors/espressif/boards/esp32/ports/common_io/iot_i2c.c#L141)
> Problem 2:
> There seems to be a bug in the iot_i2c_read_sync function with the following "!!" that I believe should only be "!" when checking for !iot_i2c_handler->is_slave_addr_write_set
> 
There is a bug and will fix it.
> Problem 3:
> We can see that some of the capabilities of common_io allow for: eI2CGetTxNoOfbytes, however from my understanding ESP32 does not have this kind of capability. As such, if someone is trying to create a portable library with common_io in mind, how would he go about knowing if eI2CGetTxNoOfbytes is supported or not and thus knowing to implement the check or not ?

ESP32 does not have support to get eI2cGetTxNoOfbytes, however we are working on a workaround to implement eI2cGetTxNoOfbytes and eI2cGetRxNoOfbytes in common_io.

We will fix these issues and update the PR at the earliest.

I apologize for late response, as I didn't receive the notification for your comment on this PR.

**Comment by shubhamkulkarni97** [comment]
> Forgot to add that I've been trying the above code and comparing direct implementation and common_io to read 1 register off of an I2C device connected to my ESP32 and the common_io version does not work.
> I believe because of the following missing part.
> 
> Also wanting to add, that am not sure how the io_ctl for eI2CSetSlaveAddr works in specifying/setting the following 2 flags separately:
> bool is_slave_addr_write_set;
> bool is_slave_addr_read_set;

We will also fix this issue

**Comment by teuteuguy** [comment]
Have just tried latest implementation from commit 1bc9323.

I could not get the I2C to work using the latest code. Right now, I'm just trying to read a register.

When using espressif code to read a byte I do:
`
i2c_cmd_handle_t cmd = i2c_cmd_link_create();
i2c_master_start(cmd);
i2c_master_write_byte(cmd, (device_addr << 1) | I2C_MASTER_WRITE, ACK_CHECK_EN);
i2c_master_write_byte(cmd, register_addr, ACK_CHECK_EN);
i2c_master_start(cmd);
i2c_master_write_byte(cmd, (device_addr << 1) | I2C_MASTER_READ, ACK_CHECK_EN);
if (length > 1) 
{
    i2c_master_read(cmd, rd_buff, length - 1, ACK_VAL);
}
i2c_master_read_byte(cmd, rd_buff + length - 1, NACK_VAL);
i2c_master_stop(cmd);
e = i2c_master_cmd_begin(wire->i2cnum, cmd, 250 / portTICK_RATE_MS);
i2c_cmd_link_delete(cmd);
`

The iot_i2c.h file describes the flow of reading in master mode as the following:
`
// Set slave address.
lRetVal = iot_i2c_ioctl( xI2CHandle, eI2CSetSlaveAddr, &uctestIotI2CSlaveAddr );
// Write the register address as single byte, in a transaction.
lRetVal = iot_i2c_write_sync( xI2CHandle, &xDeviceRegisterAddress, sizeof( xDeviceRegisterAddress ) );
if ( lRetVal == IOT_I2C_SUCCESS )
{
    // Get the number of written bytes in last transaction.
    lRetVal = iot_i2c_ioctl( xI2CHandle, eI2CGetTxNoOfbytes, &usWriteBytes );
    // Read two bytes of data to allocated buffer, in a transaction.
    lRetVal = iot_i2c_read_sync( xI2CHandle, &ucReadBuffer, sizeof( ucReadBuffer ) );
    if ( lRetVal == IOT_I2C_SUCCESS )
    {
        // Get the number of read bytes in last transaction.
        lRetVal = iot_i2c_ioctl( xI2CHandle, eI2CGetRxNoOfbytes, &usReadBytes );
    }
}
`

If i purely transpose all these command line calls the "espressif" code will translate to:

`
i2c_ctx->slave_addr = (*(uint8_t *)pvBuffer);
i2c_ctx->is_send_start_flag_set = true;
i2c_ctx->is_slave_addr_set = true;

i2c_cmd_handle_t cmd = i2c_cmd_link_create();
ret  = i2c_master_start(cmd);
ret |= i2c_master_write_byte(cmd, iot_i2c_handler->slave_addr << 1 | I2C_MASTER_WRITE, ACK_CHECK_EN);
ret = i2c_master_write(cmd, src_buf, xBytes, ACK_CHECK_EN);
ret |= i2c_master_stop(cmd);
ret |= i2c_master_cmd_begin(i2c_port_num, cmd, 1000 / portTICK_RATE_MS );
i2c_cmd_link_delete(cmd);
iot_i2c_handler->bytes_to_write = xBytes;
iot_i2c_handler->is_send_no_stop_flag_set = false;

i2c_cmd_handle_t cmd = i2c_cmd_link_create();
// Missing the cmd_start here because in the previous write_sync, the flag gets set to false.
if (xBytes > 1) {
    ret |= i2c_master_read(cmd, src_buf, xBytes - 1, ACK_VAL);
}
ret |= i2c_master_read_byte(cmd, (uint8_t *)(src_buf + xBytes - 1), NACK_VAL);
ret |= i2c_master_cmd_begin(i2c_port_num, cmd, 1000 / portTICK_RATE_MS);
i2c_cmd_link_delete(cmd);
iot_i2c_handler->is_send_no_stop_flag_set = false;
iot_i2c_handler->bytes_to_read = xBytes;
`

Where I believe maybe a cmd_start is missing before the read portion ?

**Comment by teuteuguy** [comment]
Have just tried latest implementation from commit 1bc9323.

I could not get the I2C to work using the latest code. Right now, I'm just trying to read a register.

When using espressif code to read a byte I do:

```
i2c_cmd_handle_t cmd = i2c_cmd_link_create();
i2c_master_start(cmd);
i2c_master_write_byte(cmd, (device_addr << 1) | I2C_MASTER_WRITE, ACK_CHECK_EN);
i2c_master_write_byte(cmd, register_addr, ACK_CHECK_EN);
i2c_master_start(cmd);
i2c_master_write_byte(cmd, (device_addr << 1) | I2C_MASTER_READ, ACK_CHECK_EN);
if (length > 1) 
{
    i2c_master_read(cmd, rd_buff, length - 1, ACK_VAL);
}
i2c_master_read_byte(cmd, rd_buff + length - 1, NACK_VAL);
i2c_master_stop(cmd);
e = i2c_master_cmd_begin(wire->i2cnum, cmd, 250 / portTICK_RATE_MS);
i2c_cmd_link_delete(cmd);
```

The iot_i2c.h file describes the flow of reading in master mode as the following:
```
// Set slave address.
lRetVal = iot_i2c_ioctl( xI2CHandle, eI2CSetSlaveAddr, &uctestIotI2CSlaveAddr );
// Write the register address as single byte, in a transaction.
lRetVal = iot_i2c_write_sync( xI2CHandle, &xDeviceRegisterAddress, sizeof( xDeviceRegisterAddress ) );
if ( lRetVal == IOT_I2C_SUCCESS )
{
    // Get the number of written bytes in last transaction.
    lRetVal = iot_i2c_ioctl( xI2CHandle, eI2CGetTxNoOfbytes, &usWriteBytes );
    // Read two bytes of data to allocated buffer, in a transaction.
    lRetVal = iot_i2c_read_sync( xI2CHandle, &ucReadBuffer, sizeof( ucReadBuffer ) );
    if ( lRetVal == IOT_I2C_SUCCESS )
    {
        // Get the number of read bytes in last transaction.
        lRetVal = iot_i2c_ioctl( xI2CHandle, eI2CGetRxNoOfbytes, &usReadBytes );
    }
}
```

If i purely transpose all these command line calls the "espressif" code will translate to:

```
i2c_ctx->slave_addr = (*(uint8_t *)pvBuffer);
i2c_ctx->is_send_start_flag_set = true;
i2c_ctx->is_slave_addr_set = true;

i2c_cmd_handle_t cmd = i2c_cmd_link_create();
// is_send_start_flag_set == true
ret  = i2c_master_start(cmd);
ret |= i2c_master_write_byte(cmd, iot_i2c_handler->slave_addr << 1 | I2C_MASTER_WRITE, ACK_CHECK_EN);
is_send_start_flag_set = false
ret = i2c_master_write(cmd, src_buf, xBytes, ACK_CHECK_EN);
ret |= i2c_master_stop(cmd);
ret |= i2c_master_cmd_begin(i2c_port_num, cmd, 1000 / portTICK_RATE_MS );
i2c_cmd_link_delete(cmd);
iot_i2c_handler->bytes_to_write = xBytes;
iot_i2c_handler->is_send_no_stop_flag_set = false;

i2c_cmd_handle_t cmd = i2c_cmd_link_create();
// is_send_start_flag_set == false => skipping below
// ret  = i2c_master_start(cmd);
// ret |= i2c_master_write_byte(cmd, iot_i2c_handler->slave_addr << 1 | I2C_MASTER_READ, ACK_CHECK_EN);
// iot_i2c_handler->is_send_start_flag_set = false;
// note: Missing the cmd_start here because in the previous write_sync, the flag gets set to false.
if (xBytes > 1) {
    ret |= i2c_master_read(cmd, src_buf, xBytes - 1, ACK_VAL);
}
ret |= i2c_master_read_byte(cmd, (uint8_t *)(src_buf + xBytes - 1), NACK_VAL);
ret |= i2c_master_cmd_begin(i2c_port_num, cmd, 1000 / portTICK_RATE_MS);
i2c_cmd_link_delete(cmd);
iot_i2c_handler->is_send_no_stop_flag_set = false;
iot_i2c_handler->bytes_to_read = xBytes;
```

I have tried adding the following:
```
lRetVal = iot_i2c_ioctl( xI2CHandle, eI2CSetSlaveAddr, &uctestIotI2CSlaveAddr );
```

before the call to:
```
lRetVal = iot_i2c_read_sync( xI2CHandle, &ucReadBuffer, sizeof( ucReadBuffer ) );
```

And that works.
Effectively, the read code, tests for the is_send_start_flag_set which is false at that point, and as such, can't execute the i2c_master_write_byte(cmd, iot_i2c_handler->slave_addr << 1 | I2C_MASTER_READ, ACK_CHECK_EN); command that is needed for the read to work.

**Comment by shubhamkulkarni97** [comment]
Hi @teuteuguy ,
We have referred the test cases for implementation of all the APIs.

According to the test cases,
`lRetVal = iot_i2c_ioctl( xI2CHandle, eI2CSetSlaveAddr, &uctestIotI2CSlaveAddr );` should be called to send repeated start, before calling iot_i2c_read_sync() API, as shown here: https://github.com/aws/amazon-freertos/blob/949b0331bf733acccb39bc73af2d242d337177ef/libraries/abstractions/common_io/test/test_iot_i2c.c#L1143

**Comment by teuteuguy** [comment]
Commenting out lines 349 and 353 seems to work.
Which brings question as to what is the purpose of is_send_start_flag_set ?

**Comment by shubhamkulkarni97** [comment]
> Commenting out lines 349 and 353 seems to work.
> Which brings question as to what is the purpose of is_send_start_flag_set ?

I'll check once and get back to you

**Comment by teuteuguy** [comment]
Thanks @shubhamkulkarni97 for this information.

Indeed the test cases indicate calling the lRetVal = iot_i2c_ioctl( xI2CHandle, eI2CSetSlaveAddr, &uctestIotI2CSlaveAddr ); between the write of the destination register and begining of the read.

This is not what is indicated in the iot_i2c.h file in the comments section.

@lundinc2 any comments on this ? Maybe the .h file needs updating ?

**Comment by teuteuguy** [comment]
Actually found another potential issue with the implementation.
Currently the iot_i2c_open function can only open 1 instance of the I2C at a time, because of the i2c_flag ...

What if you want to run 2 I2C ports at same time ?

**Comment by qiutongs** [comment]
> Thanks @shubhamkulkarni97 for this information.
> 
> Indeed the test cases indicate calling the lRetVal = iot_i2c_ioctl( xI2CHandle, eI2CSetSlaveAddr, &uctestIotI2CSlaveAddr ); between the write of the destination register and begining of the read.
> 
> This is not what is indicated in the iot_i2c.h file in the comments section.
> 
> @lundinc2 any comments on this ? Maybe the .h file needs updating ?



> Thanks @shubhamkulkarni97 for this information.
> 
> Indeed the test cases indicate calling the lRetVal = iot_i2c_ioctl( xI2CHandle, eI2CSetSlaveAddr, &uctestIotI2CSlaveAddr ); between the write of the destination register and begining of the read.
> 
> This is not what is indicated in the iot_i2c.h file in the comments section.
> 
> @lundinc2 any comments on this ? Maybe the .h file needs updating ?

@shubhamkulkarni97 
We apologize that the following test code and comment confused you about the expected behavior or usage of the I2C APIs. (That line can be actually deleted without affecting the test behavior and I will delete it)

https://github.com/aws/amazon-freertos/blob/949b0331bf733acccb39bc73af2d242d337177ef/libraries/abstractions/common_io/test/test_iot_i2c.c#L1142-L1143

eI2CSendNoStopFlag is designed to indicate what operation goes into one transaction while eI2CSetSlaveAddr is not. Slave address should be memorized and eI2CSetSlaveAddr is only called again when you talk to a different slave device.

Combing multiple operations in one transaction is done through eI2CSendNoStopFlag in following way:

```
// example of one transaction: write-then-read

iot_i2c_ioctl( xI2CHandle, eI2CSendNoStopFlag, NULL );
iot_i2c_write_sync(xI2CHandle, ...);
// repeated start is automatically applied here because there was "no stop" in the above write_sync
iot_i2c_read_sync(xI2CHandle, ...);
```

Above code contrasts with:
```
// example of two transactions: write and read

// stop condition is applied in the end of call iot_i2c_write_sync
iot_i2c_write_sync(xI2CHandle, ...);
// stop condition is applied in the end of call iot_i2c_read_sync
iot_i2c_read_sync(xI2CHandle, ...);
```

Hopefully that makes sense. I will try to schedule a call without to make the discussion more efficient.

**Comment by qiutongs** [comment]
> Actually found another potential issue with the implementation.
> Currently the iot_i2c_open function can only open 1 instance of the I2C at a time, because of the i2c_flag ...
> 
> What if you want to run 2 I2C ports at same time ?

Yes, I agree. The "open flag" should not be global and it should be scoped to each I2C instances. See [this example](https://github.com/aws/amazon-freertos/blob/master/vendors/st/boards/stm32l475_discovery/ports/common_io/iot_i2c.c#L84) I implemented for ST board. I used an integer to hold status of an I2C instance and "open flag" is one of the status (one bit of that integer).

File: vendors/espressif/boards/esp32/ports/common_io/iot_i2c.c [hunk 1]
```diff
@@ -0,0 +1,436 @@
...
+    IotI2CConfig_t iot_i2c_config;
+    int32_t i2c_port_num;
+    bool driver_installed;
+    void (*func)(IotI2COperationStatus_t arg1, void *arg2);
+    void *arg;
+    void *pvBuffer;
+    size_t xBytes;
+    bool is_channel_busy;
+    bool is_slave_addr_set;
+    bool is_send_start_flag_set;
```
Comment (qiutongs) [inline vendors/espressif/boards/esp32/ports/common_io/iot_i2c.c:44]: As commented, I think is_send_start_flag_set is not needed.

File: vendors/espressif/boards/esp32/ports/common_io/iot_i2c.c [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+    case eI2CSetMasterConfig : {
+        if (i2c_ctx->is_channel_busy) {
+            return IOT_I2C_BUSY;
+        }
+        IotI2CConfig_t *iot_i2c_handler = (IotI2CConfig_t *) pvBuffer;
+        i2c_ctx->iot_i2c_config.ulBusFreq = iot_i2c_handler->ulBusFreq;
+        i2c_ctx->iot_i2c_config.ulMasterTimeout = iot_i2c_handler->ulMasterTimeout;
+        int32_t i2c_port_num = (int32_t) i2c_ctx->i2c_port_num;
+
+        if (i2c_ctx->driver_installed && i2c_ctx->is_channel_busy == false) {
```
Comment (qiutongs) [inline vendors/espressif/boards/esp32/ports/common_io/iot_i2c.c:97]: Is there any reason you re-check the i2c_ctx->is_channel_busy? (It is already checked in above code)

**Comment by shubhamkulkarni97** [comment]
@qiutongs 

Thanks for the clarification, we will work on the suggestions!

File: libraries/abstractions/common_io/test/test_iot_i2c.c [hunk 3]
```diff
@@ -148,7 +148,7 @@ void prvI2CCallback( IotI2COperationStatus_t xOpStatus,
     if( xOpStatus == eI2CCompleted )
     {
         xSemaphoreGiveFromISR( xtestIotI2CSemaphore, &xHigherPriorityTaskWoken );
-        portYIELD_FROM_ISR( xHigherPriorityTaskWoken );
+        portYIELD_FROM_ISR( );
```
Comment (qiutongs) [inline libraries/abstractions/common_io/test/test_iot_i2c.c:151]: I didn't realize this is port-specific. I will figure out a plan for this.

File: libraries/abstractions/common_io/test/test_iot_uart.c [hunk 4]
```diff
@@ -63,7 +63,7 @@
 /* Globals values which can be overwritten by the test
  * framework invoking these tests */
 /*-----------------------------------------------------------*/
-uint8_t uctestIotUartPort = 0; /** Default UART port for testing */
+uint8_t uctestIotUartPort = 2; /** Default UART port for testing */
```
Comment (qiutongs) [inline libraries/abstractions/common_io/test/test_iot_uart.c:66]: Please do not modify the value here. It is supposed to be configured in ports/common_io/iot_test_common_io_internal.c. See below example for ST board.

https://github.com/aws/amazon-freertos/blob/master/vendors/st/boards/stm32l475_discovery/ports/common_io/iot_test_common_io_internal.c#L37

File: vendors/espressif/boards/esp32/aws_tests/config_files/iot_test_common_io_config.h [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+extern const uint32_t uartIotUartStopBits[ UART_TEST_SET ];
+
+/* How many I2C ports are being test. */
+#define I2C_TEST_SET                        1
+
+#define IOT_TEST_COMMON_IO_I2C_SUPPORTED                     1
+
+#if ( IOT_TEST_COMMON_IO_I2C_SUPPORTED == 1 )
+    #define IOT_TEST_COMMON_IO_I2C_SUPPORTED_SEND_NO_STOP    1
+    #define IOT_TEST_COMMON_IO_I2C_SUPPORTED_CANCEL          0
```
Comment (qiutongs) [inline vendors/espressif/boards/esp32/aws_tests/config_files/iot_test_common_io_config.h:56]: Just confirm this is done on purpose which means the cancel is not supported by I2C driver on ESP.

File: vendors/espressif/boards/esp32/ports/common_io/iot_i2c.c [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+    bool driver_installed;
+    void (*func)(IotI2COperationStatus_t arg1, void *arg2);
+    void *arg;
+    void *pvBuffer;
+    size_t xBytes;
+    bool is_channel_busy;
+    bool is_slave_addr_set;
+    bool is_send_start_flag_set;
+    bool is_send_no_stop_flag_set;
+    bool register_buffer_write_sq;
```
Comment (qiutongs) [inline vendors/espressif/boards/esp32/ports/common_io/iot_i2c.c:46]: Not seen where register_buffer_write_sq is used.

File: vendors/espressif/boards/esp32/ports/common_io/iot_i2c.c [hunk 7]
```diff
@@ -0,0 +1,436 @@
...
+{
+    if (lI2CInstance < 0 || lI2CInstance > 1) {
+        ESP_LOGE(TAG, "Invalid arguments");
+        return NULL;
+    }
+    if (i2c_flag == true) {
+        ESP_LOGE(TAG, "I2C Handler is already initialised");
+        return NULL;
+    }
+    i2c_ctx_t *i2c_ctx = (i2c_ctx_t *)calloc(1, sizeof(i2c_ctx_t));
```
Comment (qiutongs) [inline vendors/espressif/boards/esp32/ports/common_io/iot_i2c.c:65]: We suggest define two global i2c_ctx_t for each I2C instance. Thus the dynamic allocation is not needed. See below example for ST board.

https://github.com/aws/amazon-freertos/blob/master/vendors/st/boards/stm32l475_discovery/ports/common_io/iot_i2c.c#L87

File: vendors/espressif/boards/esp32/ports/common_io/iot_i2c.c [hunk 8]
```diff
@@ -0,0 +1,436 @@
...
+    if (i2c_ctx == NULL) {
+        ESP_LOGE(TAG, "Could not allocate memory for i2c context");
+        return NULL;
+    }
+    IotI2CHandle_t iot_i2c_handler = (void *) i2c_ctx;
+    i2c_ctx->i2c_port_num = lI2CInstance;
+    i2c_flag = true;
```
Comment (qiutongs) [inline vendors/espressif/boards/esp32/ports/common_io/iot_i2c.c:72]: As mentioned, this is supposed to be associated with each I2C instance.

File: libraries/abstractions/common_io/test/test_iot_uart.c (hunk 4 above)
Comment (lundinc2) [reply libraries/abstractions/common_io/test/test_iot_uart.c:66]: That was from my PR, I neglected to revert this change when I was debugging. My apologies.

File: vendors/espressif/boards/esp32/ports/common_io/iot_i2c.c [hunk 9]
```diff
@@ -0,0 +1,436 @@
...
+        return IOT_I2C_SUCCESS;
+    }
+    case eI2CSendNoStopFlag : {
+        i2c_ctx->is_send_no_stop_flag_set = true;
+        return IOT_I2C_SUCCESS;
+    }
+    default :
+        ESP_LOGE(TAG, "Invalid argument");
+        return IOT_I2C_INVALID_VALUE;
+        break;
```
Comment (qiutongs) [inline vendors/espressif/boards/esp32/ports/common_io/iot_i2c.c:173]: Nitpick: remove

File: vendors/espressif/boards/esp32/ports/common_io/iot_i2c.c [hunk 10]
```diff
@@ -0,0 +1,436 @@
...
+    if (!iot_i2c_handler->is_slave_addr_set) {
+        ESP_LOGE(TAG, "Slave address not set");
+        return IOT_I2C_SLAVE_ADDRESS_NOT_SET;
+    }
+
+    if (!iot_i2c_handler->driver_installed) {
+        return IOT_I2C_INVALID_VALUE;
+    }
+
+    if (iot_i2c_handler->is_channel_busy || iot_i2c_handler->is_channel_busy) {
```
Comment (qiutongs) [inline vendors/espressif/boards/esp32/ports/common_io/iot_i2c.c:292]: I think they are two duplicate condition.

File: vendors/espressif/boards/esp32/ports/common_io/iot_test_common_io_internal.c [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+const uint32_t uartIotUartWordLength[ UART_TEST_SET ] = { 0 };
+const uint32_t uartIotUartStopBits[ UART_TEST_SET ] = { 0 };
+
+void SET_TEST_IOT_UART_CONFIG( int testSet )
+{
+    uctestIotUartPort = uartTestPort[ testSet ];
+}
+
+/* I2C */
+const uint8_t i2cTestSlaveAddr[ I2C_TEST_SET ] = { 0xD4 };
```
Comment (qiutongs) [inline vendors/espressif/boards/esp32/ports/common_io/iot_test_common_io_internal.c:47]: Could I know which slave device this is?

**Comment by teuteuguy** [comment]
I've played around with latest release and made some slight changes in order to allow for it to work.
I've managed to get I2C 0 and 1 to work in parallel with the hacks I made.

However, now I'm confronted with the following issue, my I2C settings are:
```
#define ESP_I2C_PIN_MAP  {              \
    /* I2C 0 */                         \
    {                                   \
        GPIO_NUM_22,  /* SCL */         \
        GPIO_NUM_21   /* SCA */         \
    },                                  \
    /* I2C 1 */                         \
    {                                   \
        GPIO_NUM_26,  /* SCL */         \
        GPIO_NUM_0    /* SCA */         \
    }                                   \
}
```

When I run WITHOUT BLE, it works. 
When I try to run WITH BLE (for wifi cred provisioning) I2C 1 fails.

Is this normal ?

**Comment by shubhamkulkarni97** [comment]
> I've played around with latest release and made some slight changes in order to allow for it to work.
> I've managed to get I2C 0 and 1 to work in parallel with the hacks I made.
> 
> However, now I'm confronted with the following issue, my I2C settings are:
> 
> ```
> #define ESP_I2C_PIN_MAP  {              \
>     /* I2C 0 */                         \
>     {                                   \
>         GPIO_NUM_22,  /* SCL */         \
>         GPIO_NUM_21   /* SCA */         \
>     },                                  \
>     /* I2C 1 */                         \
>     {                                   \
>         GPIO_NUM_26,  /* SCL */         \
>         GPIO_NUM_0    /* SCA */         \
>     }                                   \
> }
> ```
> 
> When I run WITHOUT BLE, it works.
> When I try to run WITH BLE (for wifi cred provisioning) I2C 1 fails.
> 
> Is this normal ?

@teuteuguy This should not happen ideally, we try to reproduce this issue after updating I2C implementation.

**Comment by teuteuguy** [comment]
Hi @shubhamkulkarni97 thanks for confirming.

You can find my code here: https://github.com/teuteuguy/amazon-freertos-workshop/tree/feature/bmp280

If this helps.

File: vendors/espressif/boards/esp32/ports/common_io/iot_i2c.c (hunk 1 above)
Comment (shubhamkulkarni97) [reply vendors/espressif/boards/esp32/ports/common_io/iot_i2c.c]: Removed the start flag as discussed

File: vendors/espressif/boards/esp32/ports/common_io/iot_i2c.c (hunk 2 above)
Comment (shubhamkulkarni97) [reply vendors/espressif/boards/esp32/ports/common_io/iot_i2c.c]: Removed the check

File: vendors/espressif/boards/esp32/aws_tests/config_files/iot_test_common_io_config.h (hunk 5 above)
Comment (shubhamkulkarni97) [reply vendors/espressif/boards/esp32/aws_tests/config_files/iot_test_common_io_config.h:56]: Yes, cancel is not supported by I2C driver on ESP

File: vendors/espressif/boards/esp32/ports/common_io/iot_i2c.c (hunk 8 above)
Comment (shubhamkulkarni97) [reply vendors/espressif/boards/esp32/ports/common_io/iot_i2c.c]: Modified the code to use two bits, each for a I2C instance

File: vendors/espressif/boards/esp32/ports/common_io/iot_test_common_io_internal.c (hunk 11 above)
Comment (shubhamkulkarni97) [reply vendors/espressif/boards/esp32/ports/common_io/iot_test_common_io_internal.c:47]: This change is from the PR @lundinc2 had raised, however I'm using DS3231 RTC for testing(Slave address is 0x68).

**Comment by teuteuguy** [comment]
Hi @shubhamkulkarni97,

I've tested the last commit, ef46583... 
I can confirm that with my code, I2C 0 and I2C 1 work in parallel now, as well as the BLE stack.

I can't comment to the full tests part of the testing process, but in regards to my functional code, I can say that I2C works well !

Great !

Thanks

**Comment by shubhamkulkarni97** [comment]
Hi @teuteuguy, Thanks for confirming.

File: vendors/espressif/boards/esp32/ports/common_io/iot_test_common_io_internal.c (hunk 11 above)
Comment (qiutongs) [reply vendors/espressif/boards/esp32/ports/common_io/iot_test_common_io_internal.c:47]: Sure. Use whatever slave device that works for you guys for now.

**Comment by qiutongs** [comment]
What tests have you run so far? I assume you test I2C and UART based on the code in iot_test_common_io_internal.c.

Have you test SPI?

**Comment by shubhamkulkarni97** [comment]
> What tests have you run so far? I assume you test I2C and UART based on the code in iot_test_common_io_internal.c.
> 
> Have you test SPI?

We have run I2C and UART test as of now.

**Comment by qiutongs** [comment]
> > What tests have you run so far? I assume you test I2C and UART based on the code in iot_test_common_io_internal.c.
> > Have you test SPI?
> 
> We have run I2C and UART test as of now.

Thanks for confirmation.

**Comment by shubhamkulkarni97** [comment]
Closing this PR as discussed

**Pull request closed without merging**